File: ios/VozOSDemo/VozOSDemo/Sections/Sonix/MidiSectionSimplified.swift
import SwiftUI
import VozOS

/// Simplified MIDI Synthesis Section using SonixMidiSynthesizer.create().
///
/// Demonstrates SonixMidiSynthesizer.create() - the zero-config factory method.
/// Compare this to MidiSection.swift which uses the Builder API for advanced configuration.
struct MidiSectionSimplified: View {
    @State private var status = "Ready"
    @State private var isSynthesizing = false
    @State private var outputFile: URL?
    @State private var player: SonixPlayer?
    @State private var isPlaying = false
    @State private var synthesizerVersion: String?

    // Observer task for cancellation
    @State private var playingTask: Task<Void, Never>?

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            Text("MIDI Synthesis (SonixMidiSynthesizer.create)")
                .font(.headline)

            Text("Status: \(status)")
                .font(.caption)
                .foregroundColor(.secondary)

            // Show FluidSynth version if available
            if let version = synthesizerVersion {
                Text("FluidSynth: \(version)")
                    .font(.caption)
                    .foregroundColor(.secondary)
            }

            // C major scale notes
            Text("Notes: C4 - D4 - E4 - F4 - G4 - A4 - B4 - C5")
                .font(.caption)
                .foregroundColor(.secondary)

            // Synthesize button
            Button(isSynthesizing ? "Synthesizing..." : "Synthesize") {
                synthesize()
            }
            .disabled(isSynthesizing)
            .buttonStyle(.borderedProminent)

            // Playback controls
            HStack(spacing: 12) {
                Button("Play") {
                    player?.play()
                }
                .disabled(outputFile == nil || isSynthesizing || isPlaying)
                .buttonStyle(.bordered)

                Button("Stop") {
                    player?.stop()
                }
                .disabled(!isPlaying)
                .buttonStyle(.bordered)
            }
        }
        .onDisappear {
            playingTask?.cancel()
            player?.release()
        }
    }

    private func synthesize() {
        Task {
            isSynthesizing = true
            status = "Synthesizing..."

            do {
                // Copy soundfont to file
                guard let sfPath = copyAssetToFile(name: "harmonium", ext: "sf2") else {
                    status = "Soundfont not found"
                    isSynthesizing = false
                    return
                }

                // C major scale (timing in milliseconds)
                let notes: [MidiNote] = [
                    MidiNote(note: 60, startTime: 0, endTime: 400),     // C4
                    MidiNote(note: 62, startTime: 500, endTime: 900),   // D4
                    MidiNote(note: 64, startTime: 1000, endTime: 1400), // E4
                    MidiNote(note: 65, startTime: 1500, endTime: 1900), // F4
                    MidiNote(note: 67, startTime: 2000, endTime: 2400), // G4
                    MidiNote(note: 69, startTime: 2500, endTime: 2900), // A4
                    MidiNote(note: 71, startTime: 3000, endTime: 3400), // B4
                    MidiNote(note: 72, startTime: 3500, endTime: 4400), // C5
                ]

                let tempDir = FileManager.default.temporaryDirectory
                let outFile = tempDir.appendingPathComponent("midi_output.wav")

                // SIMPLE API: SonixMidiSynthesizer.create()
                let synth = SonixMidiSynthesizer.create(soundFontPath: sfPath)

                synthesizerVersion = synth.version

                let success = synth.synthesizeFromNotes(
                    notes: notes,
                    outputPath: outFile.path
                )

                if success {
                    let attrs = try FileManager.default.attributesOfItem(atPath: outFile.path)
                    let fileSize = (attrs[.size] as? Int64) ?? 0
                    outputFile = outFile
                    status = "Generated: \(fileSize / 1024) KB"

                    // Load into SonixPlayer for playback
                    await loadPlayer(path: outFile.path)
                } else {
                    status = "Synthesis failed"
                }
            } catch {
                status = "Error: \(error.localizedDescription)"
            }

            isSynthesizing = false
        }
    }

    private func loadPlayer(path: String) async {
        // Release old player
        playingTask?.cancel()
        player?.release()

        do {
            let newPlayer = try await SonixPlayer.create(source: path)
            await MainActor.run {
                player = newPlayer
                // Observe playback state
                playingTask = newPlayer.observeIsPlaying { self.isPlaying = $0 }
            }
        } catch {
            await MainActor.run {
                status = "Player error: \(error.localizedDescription)"
            }
        }
    }
}

#Preview {
    MidiSectionSimplified()
        .padding()
}

File: demo-apps/android/app/src/main/kotlin/com/musicmuni/voxatrace/demo/sonix/simplified/MidiSectionSimplified.kt
package com.musicmuni.voxatrace.demo.sonix.simplified

import android.content.Context
import androidx.compose.foundation.layout.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import com.musicmuni.voxatrace.sonix.SonixMidiSynthesizer
import com.musicmuni.voxatrace.sonix.SonixPlayer
import com.musicmuni.voxatrace.sonix.midi.MidiNote
import com.musicmuni.voxatrace.demo.util.copyAssetToFile
import io.github.aakira.napier.Napier
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.File

/**
 * Simplified MIDI Synthesis Section using SonixMidiSynthesizer.
 *
 * Demonstrates SonixMidiSynthesizer.create() - the zero-config factory method.
 */
@Composable
fun MidiSectionSimplified(context: Context) {
    val scope = rememberCoroutineScope()
    var status by remember { mutableStateOf("Ready") }
    var isSynthesizing by remember { mutableStateOf(false) }
    var outputFile by remember { mutableStateOf<File?>(null) }
    var player by remember { mutableStateOf<SonixPlayer?>(null) }
    var synthesizerVersion by remember { mutableStateOf<String?>(null) }

    // Observe player state
    val isPlaying by player?.isPlaying?.collectAsState() ?: remember { mutableStateOf(false) }

    // Cleanup
    DisposableEffect(Unit) {
        onDispose {
            player?.release()
        }
    }

    Column(
        modifier = Modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(8.dp)
    ) {
        Text(
            text = "MIDI Synthesis (SonixMidiSynthesizer.create)",
            style = MaterialTheme.typography.titleMedium
        )

        Text(text = "Status: $status", style = MaterialTheme.typography.bodySmall)

        // Show FluidSynth version if available
        synthesizerVersion?.let {
            Text(text = "FluidSynth: $it", style = MaterialTheme.typography.bodySmall)
        }

        // C major scale notes
        Text(
            text = "Notes: C4 - D4 - E4 - F4 - G4 - A4 - B4 - C5",
            style = MaterialTheme.typography.bodySmall
        )

        // Synthesize button
        Button(
            onClick = {
                scope.launch {
                    isSynthesizing = true
                    status = "Synthesizing..."

                    try {
                        // Copy soundfont to file
                        val sfFile = copyAssetToFile(context, "harmonium.sf2")

                        // C major scale
                        val notes = listOf(
                            MidiNote(note = 60, startTime = 0f, endTime = 400f),     // C4
                            MidiNote(note = 62, startTime = 500f, endTime = 900f),   // D4
                            MidiNote(note = 64, startTime = 1000f, endTime = 1400f), // E4
                            MidiNote(note = 65, startTime = 1500f, endTime = 1900f), // F4
                            MidiNote(note = 67, startTime = 2000f, endTime = 2400f), // G4
                            MidiNote(note = 69, startTime = 2500f, endTime = 2900f), // A4
                            MidiNote(note = 71, startTime = 3000f, endTime = 3400f), // B4
                            MidiNote(note = 72, startTime = 3500f, endTime = 4400f), // C5
                        )

                        val outFile = File(context.cacheDir, "midi_output.wav")

                        val (success, version) = withContext(Dispatchers.IO) {
                            // SIMPLE API: SonixMidiSynthesizer.create()
                            val synth = SonixMidiSynthesizer.create(
                                soundFontPath = sfFile.absolutePath
                            )
                            Napier.d("MIDI Synth version: ${synth.version}")

                            val result = synth.synthesizeFromNotes(
                                notes = notes,
                                outputPath = outFile.absolutePath
                            )
                            Pair(result, synth.version)
                        }

                        // Check if still active before updating UI state
                        if (!isActive) return@launch

                        synthesizerVersion = version
                        if (success && outFile.exists() && outFile.length() > 0) {
                            outputFile = outFile
                            status = "Generated: ${outFile.length() / 1024} KB"

                            // Load into SonixPlayer for playback
                            player?.release()
                            val newPlayer = SonixPlayer.create(outFile.absolutePath)
                            player = newPlayer
                        } else {
                            status = "Synthesis failed"
                        }
                    } catch (e: Exception) {
                        Napier.e("MIDI synthesis failed", e)
                        if (isActive) status = "Error: ${e.message}"
                    } finally {
                        if (isActive) isSynthesizing = false
                    }
                }
            },
            enabled = !isSynthesizing
        ) {
            Text(if (isSynthesizing) "Synthesizing..." else "Synthesize")
        }

        // Playback controls
        Row(
            horizontalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            Button(
                onClick = { player?.play() },
                enabled = outputFile != null && !isSynthesizing && !isPlaying
            ) {
                Text("Play")
            }

            Button(
                onClick = { player?.stop() },
                enabled = isPlaying
            ) {
                Text("Stop")
            }
        }
    }
}
